// @strapi/design-system/dist/Select/SelectButton.js
import u, { forwardRef as b } from "react";
import e from "prop-types";
import c from "styled-components";
import { UpState as f, DownState as y } from "./constants.js";
import { KeyboardKeys as t } from "../helpers/keyboardKeys.js";
const m = c.button`
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  top: 0;
  width: 100%;
  background: transparent;
  border: none;

  // The focus state is moved to the parent thanks to :focus-within
  &:focus {
    outline: none;
  }

  &[aria-disabled='true'] {
    cursor: not-allowed;
  }
`, a = b(({ labelledBy: s, expanded: d, onTrigger: r, disabled: n, ...l }, i) => {
  const p = (o) => {
    if (!n)
      switch (o.key) {
        case t.DOWN:
        case t.SPACE:
        case t.ENTER: {
          o.preventDefault(), r(y.Keyboard);
          break;
        }
        case t.UP: {
          o.preventDefault(), r(f.Keyboard);
          break;
        }
      }
  };
  return /* @__PURE__ */ u.createElement(
    m,
    {
      ref: i,
      "aria-labelledby": s,
      "aria-haspopup": "listbox",
      "aria-expanded": d,
      onKeyDown: p,
      "aria-disabled": n,
      type: "button",
      ...l
    }
  );
});
a.displayName = "SelectButton";
a.defaultProps = {
  expanded: !1,
  disabled: !1
};
a.propTypes = {
  disabled: e.bool,
  expanded: e.bool,
  labelledBy: e.string.isRequired,
  onTrigger: e.func.isRequired
};
export {
  a as SelectButton
};
