// @strapi/design-system/dist/v2/index.js
import { Breadcrumbs as e } from "./Breadcrumbs/Breadcrumbs.js";
import { Crumb as m } from "./Breadcrumbs/Crumb.js";
import { CrumbLink as n } from "./Breadcrumbs/CrumbLink.js";
import { CrumbSimpleMenu as f } from "./Breadcrumbs/CrumbSimpleMenu.js";
import { Link as i } from "./Link/Link.js";
import { LinkButton as v } from "./LinkButton/LinkButton.js";
import { MainNav as S } from "./MainNav/MainNav.js";
import { NavBrand as k } from "./MainNav/NavBrand.js";
import { NavCondense as s } from "./MainNav/NavCondense.js";
import { NavFooter as d } from "./MainNav/NavFooter.js";
import { NavLink as M } from "./MainNav/NavLink.js";
import { NavSection as P } from "./MainNav/NavSection.js";
import { NavSections as l } from "./MainNav/NavSections.js";
import { NavUser as D } from "./MainNav/NavUser.js";
import { Pagination as H } from "./Pagination/Pagination.js";
import { Dots as U, NextLink as W, PageLink as j, PreviousLink as q } from "./Pagination/components.js";
import { MenuItem as y, SimpleMenu as z } from "./SimpleMenu/SimpleMenu.js";
import { SubNav as E, subNavWidth as G } from "./SubNav/SubNav.js";
import { SubNavHeader as K } from "./SubNav/SubNavHeader.js";
import { SubNavLink as Q } from "./SubNav/SubNavLink.js";
import { SubNavLinkSection as T } from "./SubNav/SubNavLinkSection.js";
import { SubNavSection as X } from "./SubNav/SubNavSection.js";
import { SubNavSections as Z } from "./SubNav/SubNavSections.js";
export {
  e as Breadcrumbs,
  m as Crumb,
  n as CrumbLink,
  f as CrumbSimpleMenu,
  U as Dots,
  i as Link,
  v as LinkButton,
  S as MainNav,
  y as MenuItem,
  k as NavBrand,
  s as NavCondense,
  d as NavFooter,
  M as NavLink,
  P as NavSection,
  l as NavSections,
  D as NavUser,
  W as NextLink,
  j as PageLink,
  H as Pagination,
  q as PreviousLink,
  z as SimpleMenu,
  E as SubNav,
  K as SubNavHeader,
  Q as SubNavLink,
  T as SubNavLinkSection,
  X as SubNavSection,
  Z as SubNavSections,
  G as subNavWidth
};
