// @strapi/design-system/dist/Combobox/components.js
import n from "styled-components";
import { Box as i } from "../Box/Box.js";
import { Flex as e } from "../Flex/Flex.js";
import { inputFocusStyle as t } from "../themes/utils.js";
const c = n(e)`
  position: relative;
  border: 1px solid ${({ theme: o, hasError: r }) => r ? o.colors.danger600 : o.colors.neutral200};
  padding-right: ${({ theme: o }) => o.spaces[3]};
  padding-left: ${({ theme: o }) => o.spaces[3]};
  border-radius: ${({ theme: o }) => o.borderRadius};
  background: ${({ theme: o }) => o.colors.neutral0};

  ${({ theme: o, $disabled: r }) => r ? `
    color: ${o.colors.neutral600};
    background: ${o.colors.neutral150};
  ` : void 0}

  ${t()}
`, p = n.div`
  padding: 1px 2px;
  grid-area: 1 / 1 / 2 / 3;
`, u = n(e)`
  display: grid;
  flex: 1 1 0%;
  position: relative;
`, g = n.input`
  display: inline-grid;
  grid-area: 1 / 1 / 2 / 3;
  grid-template-columns: 0px min-content;
  background: transparent;
  min-height: ${40 / 16}rem;
  border: none;
  flex: 1;
  font-size: ${14 / 16}rem;
  color: ${({ theme: o }) => o.colors.neutral800};
  outline: none;
  &:focus-visible {
    outline: none;
    box-shadow: none;
    outline-offset: 0;
  }
  &[aria-disabled='true'] {
    background: inherit;
    color: inherit;
    cursor: not-allowed;
  }
`, $ = n(i)`
  width: 100%;
  border: none;
  text-align: left;
  outline-offset: -3px;
  ${({ isSelected: o, theme: r }) => o && `background: ${r.colors.primary100};`}

  &:hover {
    background: ${({ theme: o }) => o.colors.primary100};
  }
`;
export {
  g as Input,
  u as InputContainer,
  c as MainRow,
  $ as OptionBox,
  p as ValueContainer
};
