// @strapi/design-system/dist/DatePicker/components.cjs
"use strict";Object.defineProperties(exports,{__esModule:{value:!0},[Symbol.toStringTag]:{value:"Module"}});const t=require("styled-components"),a=require("../Box/Box.cjs"),n=require("../Field/FieldInput.cjs"),i=require("../Popover/Popover.cjs"),d=e=>e&&typeof e=="object"&&"default"in e?e:{default:e},o=d(t),s=o.default(i.Popover)`
  max-height: ${3*6}rem;
  overflow: hidden;
`,l=o.default.button`
  border: none;
  background: transparent;
  border-radius: ${({theme:e})=>e.borderRadius};
  cursor: ${e=>e["aria-disabled"]?"not-allowed":void 0};

  & svg path {
    fill: ${({theme:e})=>e.colors.neutral500};
  }
`,c=o.default.div`
  ${({bold:e,theme:r})=>e?`& ${n.InputWrapper} {
  border: 1px solid ${r.colors.primary600};
}`:void 0}
`,u=o.default(a.Box)`
  background: transparent;
  border: none;
  position: relative;
  z-index: 1;

  svg {
    height: ${11/16}rem;
    width: ${11/16}rem;
  }

  svg path {
    fill: ${({theme:e})=>e.colors.neutral600};
  }
`;exports.DatePickerButton=l;exports.DatePickerPopover=s;exports.DatePickerWrapper=c;exports.IconBox=u;
