// @strapi/design-system/dist/Field/FieldLabel.js
import { jsx as r, jsxs as c } from "react/jsx-runtime";
import { forwardRef as h } from "react";
import t from "styled-components";
import { useField as f } from "./FieldContext.js";
import { Flex as n } from "../Flex/Flex.js";
import { once as g } from "../helpers/deprecations.js";
import { Typography as l } from "../Typography/Typography.js";
const u = g(console.warn), T = h(({
  children: e,
  action: o,
  required: i,
  ...s
}, d) => {
  const {
    id: a,
    required: m
  } = f(), p = m || i;
  return i !== void 0 && u('Deprecation warning: Usage of "required" prop in FieldLabel component is deprecated. This is discouraged and will be removed in the next major release. Please use the Field component to share the required prop.'), r(l, {
    ref: d,
    variant: "pi",
    textColor: "neutral800",
    htmlFor: a,
    fontWeight: "bold",
    as: "label",
    ...s,
    children: c(n, {
      alignItems: "center",
      children: [e, p && r(x, {
        textColor: "danger600",
        children: "*"
      }), o && r(F, {
        marginLeft: 1,
        children: o
      })]
    })
  });
}), x = t(l)`
  line-height: 0;
`, F = t(n)`
  line-height: 0;

  svg path {
    fill: ${({
  theme: e
}) => e.colors.neutral500};
  }
`;
export {
  T as FieldLabel
};
